File: db/src/main/java/onl/ycode/stormify/SiblingGroup.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.stormify;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import static onl.ycode.stormify.StormifyManager.stormify;

class SiblingGroup {
    static final int DEFAULT_BATCH_SIZE = 32;

    private final List<WeakReference<AutoTable>> members = new ArrayList<>();

    void add(AutoTable member) {
        members.add(new WeakReference<>(member));
    }

    synchronized void batchPopulate(AutoTable trigger) {
        if (trigger.siblingGroup != this)
            return;
        trigger.siblingGroup = null;
        List<AutoTable> toPopulate = new ArrayList<>();
        toPopulate.add(trigger);
        for (WeakReference<AutoTable> ref : members) {
            if (toPopulate.size() >= DEFAULT_BATCH_SIZE)
                break;
            AutoTable member = ref.get();
            if (member != null && member != trigger && member.siblingGroup == this) {
                member.markPopulated();
                member.siblingGroup = null;
                toPopulate.add(member);
            }
        }
        if (toPopulate.size() == 1)
            stormify().forcePopulate(trigger);
        else
            stormify().batchPopulate(toPopulate);
    }
}

File: stormify/src/commonMain/kotlin/onl/ycode/stormify/SiblingGroup.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify

import kotlinx.atomicfu.locks.SynchronizedObject
import kotlinx.atomicfu.locks.synchronized

internal class SiblingGroup {
    companion object {
        const val DEFAULT_BATCH_SIZE = 32
    }

    private val lock = SynchronizedObject()
    private val members = mutableListOf<WeakRef<AutoTable>>()

    fun add(member: AutoTable) {
        synchronized(lock) {
            members.add(WeakRef(member))
        }
    }

    fun batchPopulate(trigger: AutoTable) {
        synchronized(lock) {
            if (trigger.`!siblingGroup` !== this) return
            trigger.`!siblingGroup` = null

            val toPopulate = mutableListOf(trigger)
            val alive = mutableListOf<WeakRef<AutoTable>>()
            for (ref in members) {
                val member = ref.get()
                if (member == null) continue // dead ref — skip (prune)
                alive.add(ref)
                if (toPopulate.size >= DEFAULT_BATCH_SIZE) continue
                if (member !== trigger && member.`!siblingGroup` === this) {
                    member.markPopulated()
                    member.`!siblingGroup` = null
                    toPopulate.add(member)
                }
            }
            members.clear()
            members.addAll(alive)

            val stormify = trigger.`!stormify` ?: return
            if (toPopulate.size == 1)
                stormify.populate(trigger)
            else
                stormify.batchPopulate(toPopulate)
        }
    }
}
